tensorly/backend/core.py

```diff
@@ -0,0 +1,854 @@
+import inspect
+import importlib
+import os
+import sys
+import threading
+import types
+import warnings
+
+import numpy as np
+import scipy.linalg
+import scipy.sparse.linalg
+
+
+_STATE = threading.local()
+_DEFAULT_BACKEND = None  # changed in `_initialize_default_backend`
+_KNOWN_BACKENDS = ('numpy', 'mxnet', 'pytorch', 'tensorflow', 'cupy')
+_LOADED_BACKENDS = {}
+
+
+def _initialize_default_backend():
+    """Setup the initial Tensorly backend"""
+    default = 'numpy'
+    backend = os.environ.get('TENSORLY_BACKEND', default)
+
+    if backend not in _KNOWN_BACKENDS:
+        msg = ("TENSORLY_BACKEND should be one of {%s}, got %r. Defaulting to "
+               "'%r'") % (', '.join(map(repr, _KNOWN_BACKENDS)), backend, default)
+        warnings.warn(msg, UserWarning)
+        backend = default
+
+    set_backend(backend, make_default=True)
+
+
+class set_backend(object):
+    """Sets the backend for TensorLy.
+
+    Can be used to set the backend globally, or locally as a contextmanager.
+
+    Parameters
+    ----------
+    backend : {'numpy', 'mxnet', 'pytorch', 'tensorflow', 'cupy'}
+        The name of the backend to use. Default is 'numpy'.
+    make_default : bool, optional
+        If True, the backend will become the default backend for all threads.
+        Note that this only affects threads where the backend hasn't already
+        been explicitly set. If False (default) the backend is only set
+        contextually for the local thread.
+
+    Examples
+    --------
+    Set the backend to numpy globally for this thread:
+
+    >>> import tensorly as tl
+    >>> tl.set_backend('numpy')
+
+    Set the backend to numpy globally for all threads:
+
+    >>> tl.set_backend('numpy', make_default=True)
+
+    Set the backend to use pytorch inside a context:
+
+    >>> with tl.set_backend('pytorch'):  # doctest: +SKIP
+    ...     pass
+    """
+    def __init__(self, backend, make_default=False):
+        if not isinstance(backend, Backend):
+            if backend not in _LOADED_BACKENDS:
+                # load the backend
+                if backend in _KNOWN_BACKENDS:
+                    importlib.import_module('tensorly.backend.%s_backend' % backend)
```

Comment:
So this automatically stores the backend in thread.local().backend?
We could explicitly do `new_backend = importlib.import_module`?

Reply:
No, this just ensures the requested backend is loaded - in the imported module the backend object is added to the mapping of name -> backend. The lookup in `_LOADED_BACKENDS` is where the backend object is retrieved.

We could do `new_backend = importlib.import_module(...).backend`, but that requires that the backend object be named `backend` in the module (a weird interface). Having a mapping of strings to backend also allows tensorly backends to be defined in other projects (other projects have similar interfaces (e.g. joblib), and while this isn't needed now, it may be nice later).

Reply:
It is not very explicit though.. 

I think the most common is to explicitly import or use the module name (e.g. in Joblib, `mpickle = import_module(LOKY_PICKLER)`)

Here we could simply get the backend, and do everything (including registering it etc) in the same place.. When you explicitly get the module back it is loaded in the original way as well anyway so this shouldn't break things (e.g. if you do new_backend = importlib.import_module(...), it still imports tensorly.backend.random_backend, in addition to returning said backend).